src/db.c

```diff
@@ -305,16 +305,18 @@ robj *dbRandomKey(redisDb *db) {
 
 /* Delete a key, value, and associated expiration entry if any, from the DB */
 int dbSyncDelete(redisDb *db, robj *key) {
+    dictEntry **plink;
+    int table;
     /* Deleting an entry from the expires dict will not free the sds of
      * the key, because it is shared with the main dictionary. */
     if (dictSize(db->expires) > 0) dictDelete(db->expires,key->ptr);
```

Comment:
We need a similar change here (like what we have in the lazy freeing), remove from expires dict only after calling the module callback

Reply:
After move this, I found a bug. In the callback `RM_OpenKey` calls `lookupKeyReadWithFlags` which triggers another expire.

Reply:
Interesting, let's make sure there's a consistent test for both removal of a volatile key that's not yet expired, and one that is (being expired). 

As a solution, maybe the only way to go is prevent lazy expiry from within this callback? (not only on that one key). 
I. E. We said we don't allow modifications from within that callback, so that includes lazy expiry. 
We can set a global flag and use it in expireIfNeeded to rerun 1 and abort before deleting.

Reply:
I think maybe we should return 0 because we can access expired key.

Reply:
Yes, sorry. We should indeed return 0, since we want the module to be able to access it, and we'll delete it when the notification callback returns.